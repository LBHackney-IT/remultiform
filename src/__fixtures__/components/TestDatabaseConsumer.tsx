import React, { useContext } from "react";

import { NamedSchema, Schema } from "../../database/types";
import { DatabaseContext } from "../../database-context/DatabaseContext";

export interface TestDatabaseConsumerProps<
  DBSchema extends NamedSchema<string, number, Schema> = NamedSchema<
    string,
    number,
    Schema
  >
> {
  context: DatabaseContext<DBSchema>;
}

export const TestDatabaseConsumer: React.FunctionComponent<TestDatabaseConsumerProps> = <
  DBSchema extends NamedSchema<string, number, Schema>
>(
  props: TestDatabaseConsumerProps<DBSchema>
): JSX.Element => {
  const { context } = props;

  const database = useContext(context.context);

  if (database) {
    return <span>Database name: {database.name}</span>;
  }

  return <span>No database</span>;
};

TestDatabaseConsumer.displayName = "TestDatabaseConsumer";
