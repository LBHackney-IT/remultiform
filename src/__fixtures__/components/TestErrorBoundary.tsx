import PropTypes from "prop-types";
import React from "react";

export interface TestErrorBoundaryProps {
  children: React.ReactNode;
}

export interface TestErrorBoundaryState {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  error?: any;
}

export class TestErrorBoundary extends React.Component<
  TestErrorBoundaryProps,
  TestErrorBoundaryState
> {
  static propTypes: PropTypes.ValidationMap<TestErrorBoundaryProps> = {
    children: PropTypes.node.isRequired,
  };

  static getDerivedStateFromError(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    error: any
  ): Partial<TestErrorBoundaryState> {
    return { error };
  }

  state: TestErrorBoundaryState = {};

  render(): JSX.Element {
    const { children } = this.props;
    const { error } = this.state;

    if (error) {
      return <span>Caught error: {error.toString()}</span>;
    }

    return <>{children}</>;
  }
}
