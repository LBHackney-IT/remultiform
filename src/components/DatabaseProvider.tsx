import PropTypes from "prop-types";
import React from "react";

import { DatabaseContext } from "../helpers/DatabaseContext";

import { Database } from "../store/Database";
import { NamedSchema, Schema } from "../store/types";

/**
 * The proptypes for {@link DatabaseProvider}.
 */
export interface DatabaseProviderProps<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  /**
   * An open {@link Database} or a promise that will resolve to an one.
   *
   * Normally, you would want to pass the return value of {@link Database.open}
   * in here directly.
   */
  openDatabaseOrPromise: Database<DBSchema> | Promise<Database<DBSchema>>;

  context: DatabaseContext<DBSchema>;

  /**
   * Child components of this provider.
   *
   * Anything that needs a {@link Database} to function should be contained
   * withing this subtree. Use {@link useDatabase} with
   * {@link DatabaseProviderProps.context} to access the {@link Database} from
   * within those components.
   */
  children: React.ReactNode;
}

interface DatabaseProviderState<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  database?: Database<DBSchema>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  error?: any;
}

/**
 * A React context provider for a {@link Database}.
 *
 * This will wait for the database to finish opening on mount. Until
 * {@link DatabaseProviderProps.openDatabaseOrPromise} settles, the
 * {@link DatabaseContext} will store `undefined`, the default value.
 *
 * **Important!** You will need to ensure the database closes itself if you ever
 * need to change {@link DatabaseProviderProps.openDatabaseOrPromise}, or if you
 * unmount this component. If either of those things happen, this component
 * will **throw an exception**, (unless unmounting after an error). It is
 * therefore recommended that this component is contained within an error
 * boundary that handles exceptions gracefully, probably calling
 * {@link Database.close}.
 *
 * ```js
 * const DBContext = new DatabaseContext();
 * const databasePromise = Database.open("myDatabaseWithAnError", 1);
 *
 * class Component extends React.Component {
 *   static getDerivedStateFromError(error) {
 *     return { error };
 *   }
 *
 *   state = {};
 *
 *   render() {
 *     const { error } = this.state;
 *
 *     if (error) {
 *       return <span>Something went wrong!</span>
 *     }
 *
 *     return (
 *       <DatabaseProvider
 *         context={DBContext}
 *         openDatabaseOrPromise={databasePromise}
 *       >
 *         <span>Some content that relies on the database.</span>
 *       </DatabaseProvider>
 *   }
 * }
 * ```
 */
export class DatabaseProvider<
  DBSchema extends NamedSchema<string, number, Schema>
> extends React.Component<
  DatabaseProviderProps<DBSchema>,
  DatabaseProviderState<DBSchema>,
  never
> {
  static propTypes: PropTypes.ValidationMap<
    DatabaseProviderProps<NamedSchema<string, number, Schema>>
  > = {
    openDatabaseOrPromise: PropTypes.oneOfType([
      PropTypes.instanceOf(Database).isRequired,
      PropTypes.instanceOf<
        Promise<Database<NamedSchema<string, number, Schema>>>
      >(Promise).isRequired
    ]).isRequired,

    context: PropTypes.instanceOf(DatabaseContext).isRequired,
    children: PropTypes.node.isRequired
  };

  /**
   * @ignore
   */
  state: DatabaseProviderState<DBSchema> = {};

  private isUnmounted = true;

  /**
   * @ignore
   */
  async componentDidMount(): Promise<void> {
    this.isUnmounted = false;

    const { openDatabaseOrPromise } = this.props;

    const stateUpdate: Partial<DatabaseProviderState<DBSchema>> = {};

    try {
      const database = await openDatabaseOrPromise;

      stateUpdate.database = database;
    } catch (err) {
      stateUpdate.error = err;
    }

    if (this.isUnmounted) {
      return;
    }

    this.setState(state => ({ ...state, ...stateUpdate }));
  }

  /**
   * @ignore
   */
  componentDidUpdate(
    prevProps: Readonly<DatabaseProviderProps<DBSchema>>
  ): void {
    const { error } = this.state;

    if (error) {
      throw error;
    }

    const updatedDatabase = (Object.keys(
      prevProps
    ) as (keyof typeof prevProps)[]).some(
      propName =>
        propName === "openDatabaseOrPromise" &&
        prevProps[propName] !== this.props[propName]
    );

    if (updatedDatabase) {
      this.setState({
        error: new Error(
          "Updating the openDatabaseOrPromise prop of a DatabaseProvider is unsupported"
        )
      });
    }
  }

  /**
   * @ignore
   */
  render(): JSX.Element {
    const { openDatabaseOrPromise, context, children } = this.props;
    let { database } = this.state;

    if (!database && openDatabaseOrPromise instanceof Database) {
      database = openDatabaseOrPromise;
    }

    return (
      <context.context.Provider value={database}>
        {children}
      </context.context.Provider>
    );
  }

  /**
   * @ignore
   */
  componentWillUnmount(): void {
    this.isUnmounted = true;

    const { error } = this.state;

    if (error) {
      return;
    }

    throw new Error(
      "Unmounting a DatabaseProvider outside of an error is unsupported"
    );
  }
}
