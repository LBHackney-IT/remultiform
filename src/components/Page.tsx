import PropTypes from "prop-types";
import React from "react";

import { PageComponentWrapper } from "../helpers/PageComponentWrapper/PageComponentWrapper";
import { DatabaseContext } from "../helpers/DatabaseContext";

import { NamedSchema, Schema } from "../store/types";

/**
 * The proptypes for {@link Page}.
 */
export interface PageProps<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  /**
   * A context wrapper for a {@link Database} instance.
   *
   * You must provide this if any of your components are instances of
   * {@link DynamicPageComponent} for them to work as expected.
   */
  context?: DatabaseContext<DBSchema> | null;

  /**
   * An ordered array of wrapped components to display on the page.
   *
   * Create {@link PageComponentWrapper|PageComponentWrappers} with
   * {@link PageComponentWrapper.wrapStatic} or
   * {@link PageComponentWrapper.wrapDynamic}.
   */
  componentWrappers: PageComponentWrapper<DBSchema>[];
}

/**
 * A component for rendering a page of a multipage form.
 */
export const Page = <DBSchema extends NamedSchema<string, number, Schema>>(
  props: PageProps<DBSchema>
): JSX.Element => {
  const { context, componentWrappers } = props;

  if (context) {
    return (
      <context.Consumer>
        {(database): JSX.Element[] =>
          componentWrappers.map(({ render }) => render({ database }))
        }
      </context.Consumer>
    );
  }

  return <>{componentWrappers.map(({ render }) => render({}))}</>;
};

Page.propTypes = {
  context: PropTypes.instanceOf(DatabaseContext),
  componentWrappers: PropTypes.arrayOf(PageComponentWrapper.propType.isRequired)
    .isRequired
} as PropTypes.ValidationMap<PageProps<NamedSchema<string, number, Schema>>>;
