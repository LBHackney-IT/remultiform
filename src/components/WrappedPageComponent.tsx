import { nullAsUndefined } from "null-as-undefined";
import PropTypes from "prop-types";
import React from "react";

import {
  DynamicPageComponent,
  DynamicPageComponentControlledProps,
  DynamicPageComponentType
} from "../helpers/PageComponentWrapper/DynamicPageComponent";

import { Database } from "../store/Database";
import { NamedSchema, Schema, StoreValue, StoreNames } from "../store/types";

/**
 * The proptypes for {@link WrappedPageComponent}.
 */
export interface WrappedPageComponentProps<
  Props,
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * An open {@link Database}.
   *
   * This is usually provided by a {@link DatabaseContext.Consumer} and is
   * optional to allow waiting for {@link Database.open} to settle.
   */
  database?: Database<DBSchema> | null;

  /**
   * The {@link DynamicPageComponent} to wrap.
   */
  component: DynamicPageComponent<
    DynamicPageComponentType<Props, StoreValue<DBSchema["schema"], StoreName>>,
    Props,
    DBSchema,
    StoreName
  >;

  onChange?:
    | ((value: StoreValue<DBSchema["schema"], StoreName>) => void)
    | null;
}

interface WrappedPageComponentState<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  value?: "" | StoreValue<DBSchema["schema"], StoreName>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  error?: any;
  isFetching: boolean;
}

/**
 * A wrapped {@link DynamicPageComponent}.
 *
 * This handles interfacing between the
 * {@link DynamicPageComponentControlledProps} of a component and a
 * {@link Database} instance via a {@link DatabaseContext} and a
 * {@link DatabaseMap}.
 *
 * This component passes {@link DynamicPageComponentControlledProps.disabled}
 * as `true` when waiting for {@link Database} operations to settle. The
 * wrapped component should disable functionality appropriately to ensure data
 * doesn't get out of sync.
 */
export class WrappedPageComponent<
  Props,
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> extends React.Component<
  WrappedPageComponentProps<Props, DBSchema, StoreName>,
  WrappedPageComponentState<DBSchema, StoreName>,
  never
> {
  static propTypes: PropTypes.ValidationMap<
    WrappedPageComponentProps<
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      any,
      NamedSchema<string, number, Schema>,
      StoreNames<Schema>
    >
  > = {
    database: PropTypes.instanceOf(Database),
    component: DynamicPageComponent.propType.isRequired,
    onChange: PropTypes.func
  };

  /**
   * @ignore
   */
  state: WrappedPageComponentState<DBSchema, StoreName> = {
    value: this.props.component.emptyValue,
    isFetching: false
  };

  private isUnmounted = false;

  /**
   * @ignore
   */
  async componentDidMount(): Promise<void> {
    this.isUnmounted = false;

    // We do this here rather than in `getDerivedStateFromProps` because
    // it's asynchronous, and `getDerivedStateFromProps` is expected to
    // immediately return the new state.
    await this.fetchValue();
  }

  /**
   * @ignore
   */
  async componentDidUpdate(
    prevProps: Readonly<WrappedPageComponentProps<Props, DBSchema, StoreName>>
  ): Promise<void> {
    // We only want to load from the database if it has just opened. We
    // shouldn't be changing the underlying database while this is mounted, but
    // we don't enforce that.
    if (prevProps.database !== this.props.database) {
      // We do this here rather than in `getDerivedStateFromProps` because
      // it's asynchronous, and `getDerivedStateFromProps` is expected to
      // immediately return the new state.
      await this.fetchValue();
    }
  }

  /**
   * @ignore
   */
  render(): JSX.Element {
    const { component } = this.props;
    const { error } = this.state;

    if (error) {
      console.error(error);

      // We should make this customizable.
      return <div>Something went wrong!</div>;
    }

    const { database } = this.props;
    const { value, isFetching } = this.state;
    const { key, Component, props } = component;

    const controlledProps: DynamicPageComponentControlledProps<StoreValue<
      DBSchema["schema"],
      StoreName
    >> = {
      value,
      onValueChange: this.onValueChange.bind(this),
      disabled: !database || isFetching
    };

    return <Component key={key} {...props} {...controlledProps} />;
  }

  /**
   * @ignore
   */
  componentWillUnmount(): void {
    this.isUnmounted = true;
  }

  private async fetchValue(): Promise<void> {
    if (this.isUnmounted) {
      return;
    }

    const {
      database,
      component: { databaseMap, defaultValue, emptyValue }
    } = this.props;

    if (!database || !databaseMap) {
      return;
    }

    const { storeName, key } = databaseMap;

    this.setState(state => ({ ...state, isFetching: true }));

    const stateUpdate: Partial<WrappedPageComponentState<
      DBSchema,
      StoreName
    >> = {
      isFetching: false
    };

    try {
      // If `database.get` resolves to be undefined, this will remove the
      // current `value` from the state, if one is set.
      stateUpdate.value = await database.get(storeName, key);

      if (stateUpdate.value === undefined) {
        stateUpdate.value = nullAsUndefined(defaultValue);
      }

      if (stateUpdate.value === undefined) {
        // We need to do this to keep the component as a controlled component.
        // The controlled prop must always be defined.
        stateUpdate.value = emptyValue;
      }

      // Clear the error if it was set.
      stateUpdate.error = undefined;
    } catch (err) {
      stateUpdate.error = err;
    }

    if (this.isUnmounted) {
      return;
    }

    this.setState(state => ({ ...state, ...stateUpdate }));
  }

  private onValueChange(
    value: StoreValue<DBSchema["schema"], StoreName>
  ): void {
    if (this.isUnmounted) {
      return;
    }

    const { onChange } = this.props;

    if (onChange) {
      onChange(value);
    }

    this.setState(state => ({ ...state, value }));
  }
}
