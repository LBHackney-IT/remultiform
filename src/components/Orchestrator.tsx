import PropTypes from "prop-types";
import React from "react";

import {
  PageComponentWrapper,
  pageComponentWrapperPropType
} from "../helpers/PageComponentWrapper";

import { Page } from "./Page";

/**
 * A step in a multipage form.
 *
 * This represents a single step in the flow of the multipage form.
 *
 * ```ts
 * const step: Step = {
 *   key: "step-1",
 *   componentWrappers: [
 *     wrapPageComponent({
 *       key: "my-image",
 *       Component: "img",
 *       props: {
 *         src: "/path/to/my.png"
 *       }
 *     }),
 *     wrapPageComponent({
 *       key: "my-input",
 *       Component: MyInput,
 *       props: {
 *         defaultValue: "Enter something?"
 *       }
 *     })
 *   ]
 * };
 * ```
 */
export interface Step {
  /**
   * The key to uniquely identify the step
   *
   * This is referenced by {@link OrchestratorProps.currentStepKey}.
   */
  key: React.Key;

  /**
   * An ordered array of wrapped components to include in this step.
   *
   * Create {@link PageComponentWrapper|PageComponentWrappers} with
   * {@link wrapPageComponent}.
   */
  componentWrappers: PageComponentWrapper[];
}

/**
 * The proptypes for {@link Orchestrator}.
 */
export interface OrchestratorProps {
  /**
   * The key of the current step to be rendered.
   */
  currentStepKey?: React.Key | null;

  /**
   * An array of steps to possibly render based on
   * {@link OrchestratorProps.currentStepKey}.
   */
  steps: Step[];
}

/**
 * A component for orchestrating the rendering of pages for a multipage form.
 *
 * ```ts
 * const form: React.FunctionComponent = () => {
 *   return (
 *     <Orchestrator
 *       currentStepKey="step-2"
 *       steps={[
 *         {
 *           key: "step-1",
 *           componentWrappers: [
 *             wrapPageComponent({
 *               key: "my-image",
 *               Component: "img",
 *               props: {
 *                 src: "/path/to/my.png"
 *               }
 *             }),
 *             wrapPageComponent({
 *               key: "my-input",
 *               Component: MyInput,
 *               props: {
 *                 defaultValue: "Enter something?"
 *               }
 *             })
 *           ]
 *         },
 *         {
 *           key: "step-2",
 *           componentWrappers: [
 *             wrapPageComponent({
 *               key: "my-input",
 *               Component: MyInput,
 *               props: {
 *                 defaultValue: "Enter something else?"
 *               }
 *             })
 *           ]
 *         }
 *       ]}
 *     />
 *   )
 * }
 */
export const Orchestrator: React.FunctionComponent<OrchestratorProps> = (
  props: OrchestratorProps
): JSX.Element => {
  const { currentStepKey, steps } = props;

  const currentStep =
    steps.find(({ key }) => key === currentStepKey) || steps[0];

  return (
    <>
      <Page
        key={currentStep.key}
        componentWrappers={currentStep.componentWrappers}
      />
    </>
  );
};

Orchestrator.propTypes = {
  currentStepKey: PropTypes.string,
  steps: PropTypes.arrayOf(
    PropTypes.exact({
      key: PropTypes.oneOfType([
        PropTypes.string.isRequired,
        PropTypes.number.isRequired
      ]).isRequired,
      componentWrappers: PropTypes.arrayOf(pageComponentWrapperPropType)
        .isRequired
    }).isRequired
  ).isRequired
};
