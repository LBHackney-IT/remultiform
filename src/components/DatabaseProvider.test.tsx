import React from "react";
import { ReactTestRenderer, act, create } from "react-test-renderer";

import { TestDatabaseConsumer } from "../__fixtures__/components/TestDatabaseConsumer";

import {
  spyOnConsoleError,
  spyOnDatabaseOpen
} from "../__tests__/helpers/spies";

import { DatabaseContext } from "../helpers/DatabaseContext";

import { Database } from "../store/Database";
import { OpenOptions } from "../store/OpenOptions";
import { NamedSchema, Schema } from "../store/types";

import { DatabaseProvider } from "./DatabaseProvider";
import { TestErrorBoundary } from "../__fixtures__/components/TestErrorBoundary";

let databseOpenSpy: jest.SpyInstance<
  Promise<Database<NamedSchema<string, number, Schema>>>,
  [string, number?, OpenOptions<Schema>?]
>;

beforeEach(() => {
  databseOpenSpy = spyOnDatabaseOpen();
});

it("renders correctly with a resolved database", async () => {
  const database = await Database.open("testDBName", 1);
  const DBContext = new DatabaseContext(database);

  const component = create(
    <DatabaseProvider context={DBContext}>
      <TestDatabaseConsumer context={DBContext} />
    </DatabaseProvider>
  );

  expect(component).toMatchSnapshot();
});

it("renders correctly before the database resolves", () => {
  const databasePromise = Database.open("testDBName", 1);
  const DBContext = new DatabaseContext(databasePromise);

  const component = create(
    <DatabaseProvider context={DBContext}>
      <TestDatabaseConsumer context={DBContext} />
    </DatabaseProvider>
  );

  expect(component).toMatchSnapshot();
});

it("renders correctly after the database resolves", async () => {
  const databasePromise = Database.open("testDBName", 1);
  const DBContext = new DatabaseContext(databasePromise);

  let component: ReactTestRenderer | undefined = undefined;

  await act(async () => {
    component = create(
      <DatabaseProvider context={DBContext}>
        <TestDatabaseConsumer context={DBContext} />
      </DatabaseProvider>
    );

    await databasePromise;
  });

  expect(component).toMatchSnapshot();
});

it("throws if the database rejects", async () => {
  let resolveInternalDatabasePromise: () => void;

  const internalDatabasePromise = new Promise(resolve => {
    resolveInternalDatabasePromise = resolve;
  });

  databseOpenSpy.mockImplementation(async () => {
    await internalDatabasePromise;

    throw new Error("test error");
  });

  const databasePromise = Database.open("testDBName", 1);
  const DBContext = new DatabaseContext(databasePromise);

  let component: ReactTestRenderer | undefined = undefined;

  const consoleErrorSpy = spyOnConsoleError();

  await act(async () => {
    component = create(
      <TestErrorBoundary>
        <DatabaseProvider context={DBContext}>
          <TestDatabaseConsumer context={DBContext} />
        </DatabaseProvider>
      </TestErrorBoundary>
    );

    resolveInternalDatabasePromise();

    await internalDatabasePromise;
  });

  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();

  consoleErrorSpy.mockRestore();

  expect(component).toMatchSnapshot();
});

it("renders correctly when changing children", async () => {
  const key = "key";

  const database = await Database.open("testDBName", 1);
  const DBContext = new DatabaseContext(database);

  const Wrapper = ({
    children
  }: {
    children: React.ReactNode;
  }): JSX.Element => (
    <DatabaseProvider context={DBContext}>{children}</DatabaseProvider>
  );

  let component: ReactTestRenderer | undefined = undefined;

  act(() => {
    component = create(
      <Wrapper key={key}>
        <span>Test content</span>
      </Wrapper>
    );
  });

  act(() => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    component!.update(
      <Wrapper key={key}>
        <div>Updated test content</div>
      </Wrapper>
    );
  });

  expect(component).toMatchSnapshot();
});

it("throws when changing the `context` prop with a different database instance", async () => {
  type TestSchema = NamedSchema<"testDBName" | "newTestDBName", 1, {}>;

  const key = "key";

  const database = await Database.open<TestSchema>("testDBName", 1);
  const DBContext = new DatabaseContext(database);

  const newDatabase = await Database.open<TestSchema>("testDBName", 1);
  const NewDBContext = new DatabaseContext(newDatabase);

  const Wrapper = ({
    context
  }: {
    context: DatabaseContext<TestSchema>;
  }): JSX.Element => (
    <TestErrorBoundary>
      <DatabaseProvider context={context}>
        <span>Test content</span>
      </DatabaseProvider>
    </TestErrorBoundary>
  );

  let component: ReactTestRenderer | undefined = undefined;

  act(() => {
    component = create(<Wrapper key={key} context={DBContext} />);
  });

  const consoleErrorSpy = spyOnConsoleError();

  act(() => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    component!.update(<Wrapper key={key} context={NewDBContext} />);
  });

  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();

  consoleErrorSpy.mockRestore();

  expect(component).toMatchSnapshot();
});

it("renders correctly when changing the `context` prop with the same database instance", async () => {
  type TestSchema = NamedSchema<"testDBName" | "newTestDBName", 1, {}>;

  const key = "key";

  const database = await Database.open<TestSchema>("testDBName", 1);
  const DBContext = new DatabaseContext(database);
  const NewDBContext = new DatabaseContext(database);

  const Wrapper = ({
    context
  }: {
    context: DatabaseContext<TestSchema>;
  }): JSX.Element => (
    <DatabaseProvider context={context}>
      <span>Test content</span>
    </DatabaseProvider>
  );

  let component: ReactTestRenderer | undefined = undefined;

  act(() => {
    component = create(<Wrapper key={key} context={DBContext} />);
  });

  act(() => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    component!.update(<Wrapper key={key} context={NewDBContext} />);
  });

  expect(component).toMatchSnapshot();
});

it("throws when unmounting without an error", async () => {
  const database = await Database.open("testDBName", 1);
  const DBContext = new DatabaseContext(database);

  let component: ReactTestRenderer | undefined = undefined;

  act(() => {
    component = create(
      <TestErrorBoundary>
        <DatabaseProvider context={DBContext}>
          <span>Test content</span>
        </DatabaseProvider>
      </TestErrorBoundary>
    );
  });

  const consoleErrorSpy = spyOnConsoleError();

  act(() => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    component!.update(
      <TestErrorBoundary>
        <span>Test content</span>
      </TestErrorBoundary>
    );
  });

  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();

  consoleErrorSpy.mockRestore();

  expect(component).toMatchSnapshot();
});

it("doesn't throw when unmounting without an error with `allowUnmounting` enabled", async () => {
  const database = await Database.open("testDBName", 1);
  const DBContext = new DatabaseContext(database);

  let component: ReactTestRenderer | undefined = undefined;

  act(() => {
    component = create(
      <TestErrorBoundary>
        <DatabaseProvider context={DBContext} allowUnmounting>
          <span>Test content</span>
        </DatabaseProvider>
      </TestErrorBoundary>
    );
  });

  const consoleErrorSpy = spyOnConsoleError();

  act(() => {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    component!.update(
      <TestErrorBoundary>
        <span>Test content</span>
      </TestErrorBoundary>
    );
  });

  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();

  consoleErrorSpy.mockRestore();

  expect(component).toMatchSnapshot();
});
