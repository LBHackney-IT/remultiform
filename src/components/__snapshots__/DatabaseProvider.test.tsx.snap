// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't throw when unmounting without an error with \`allowUnmounting\` enabled 1`] = `Array []`;

exports[`doesn't throw when unmounting without an error with \`allowUnmounting\` enabled 2`] = `
<span>
  Test content
</span>
`;

exports[`renders correctly after the database resolves 1`] = `
<span>
  Database name: 
  testDBName
</span>
`;

exports[`renders correctly before the database resolves 1`] = `
<span>
  No database
</span>
`;

exports[`renders correctly when changing children 1`] = `
<div>
  Updated test content
</div>
`;

exports[`renders correctly when changing the \`context\` prop with the same database instance 1`] = `
<span>
  Test content
</span>
`;

exports[`renders correctly with a resolved database 1`] = `
<span>
  Database name: 
  testDBName
</span>
`;

exports[`throws if the database rejects 1`] = `
Array [
  Array [
    "Error: Uncaught [Error: test error]
    at reportException (<rootDir>/node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:66:24)
    at invokeEventListeners (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:209:9)
    at HTMLUnknownElementImpl._dispatch (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
    at HTMLUnknownElementImpl.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
    at HTMLUnknownElementImpl.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:30:27)
    at HTMLUnknownElement.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:157:21)
    at Object.invokeGuardedCallbackDev (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11807:16)
    at invokeGuardedCallback (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11860:31)
    at commitRootImpl (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15092:9)
    at unstable_runWithPriority (<rootDir>/node_modules/scheduler/cjs/scheduler.development.js:697:12)",
    [Error: test error],
  ],
  Array [
    "The above error occurred in the <DatabaseProvider> component:
    in DatabaseProvider
    in TestErrorBoundary

React will try to recreate this component tree from scratch using the error boundary you provided, TestErrorBoundary.",
  ],
]
`;

exports[`throws if the database rejects 2`] = `
<span>
  Caught error: 
  Error: test error
</span>
`;

exports[`throws when changing the \`context\` prop with a different database instance 1`] = `
Array [
  Array [
    "Error: Uncaught [Error: Updating the context prop of a DatabaseProvider to one with a different database is unsupported]
    at reportException (<rootDir>/node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:66:24)
    at invokeEventListeners (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:209:9)
    at HTMLUnknownElementImpl._dispatch (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
    at HTMLUnknownElementImpl.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
    at HTMLUnknownElementImpl.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:30:27)
    at HTMLUnknownElement.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:157:21)
    at Object.invokeGuardedCallbackDev (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11807:16)
    at invokeGuardedCallback (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11860:31)
    at commitRootImpl (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:15092:9)
    at unstable_runWithPriority (<rootDir>/node_modules/scheduler/cjs/scheduler.development.js:697:12)",
    [Error: Updating the context prop of a DatabaseProvider to one with a different database is unsupported],
  ],
  Array [
    "The above error occurred in the <DatabaseProvider> component:
    in DatabaseProvider (created by Wrapper)
    in TestErrorBoundary (created by Wrapper)
    in Wrapper

React will try to recreate this component tree from scratch using the error boundary you provided, TestErrorBoundary.",
  ],
]
`;

exports[`throws when changing the \`context\` prop with a different database instance 2`] = `
<span>
  Caught error: 
  Error: Updating the context prop of a DatabaseProvider to one with a different database is unsupported
</span>
`;

exports[`throws when unmounting without an error 1`] = `
Array [
  Array [
    "Error: Uncaught [Error: Unmounting a DatabaseProvider outside of an error is unsupported]
    at reportException (<rootDir>/node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:66:24)
    at invokeEventListeners (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:209:9)
    at HTMLUnknownElementImpl._dispatch (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:119:9)
    at HTMLUnknownElementImpl.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:82:17)
    at HTMLUnknownElementImpl.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:30:27)
    at HTMLUnknownElement.dispatchEvent (<rootDir>/node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:157:21)
    at Object.invokeGuardedCallbackDev (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11807:16)
    at invokeGuardedCallback (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11860:31)
    at safelyCallComponentWillUnmount (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12018:5)
    at commitUnmount (<rootDir>/node_modules/react-test-renderer/cjs/react-test-renderer.development.js:12507:11)",
    [Error: Unmounting a DatabaseProvider outside of an error is unsupported],
  ],
  Array [
    "The above error occurred in the <DatabaseProvider> component:
    in DatabaseProvider
    in TestErrorBoundary

React will try to recreate this component tree from scratch using the error boundary you provided, TestErrorBoundary.",
  ],
]
`;

exports[`throws when unmounting without an error 2`] = `
<span>
  Caught error: 
  Error: Unmounting a DatabaseProvider outside of an error is unsupported
</span>
`;
